### PR TITLE
Enable cancelling localization sessions

### PR DESCRIPTION
**Pull request by chrisfromwork**
This review addresses: https://github.com/microsoft/MixedReality-SpectatorView/issues/242

In this review:
1) ILocalizationSession is extended to have a Cancel function defined. This is beneficial because it eases cleanup when a network connection is lost.
2) SpatialLocalizationSession was created as a base class. This base class contains cancellation token sources for tearing down localization on cancel.
3) Dispatcher and DispatcherUnity were created to ease scheduling tasks on the ui thread.
4) We now lock when accessing the current localization session in SpatialCoordinateSystemManager.
5) SpatialCoordinateSystemManager now has a flag that states whether localization is running.
6) SpatialAlignmentVisual now displays to the user when localization is running.

### Notes:
This review introduces breaking changes to ISpatialLocalizationSession. Contributors felt there was a need to be able to cancel sessions, which led to this change.

### Migration Instructions:
1) Update custom implementations of ISpatialLocalizationSession to support a Cancel function.

### CODE REVIEW

File: src/SpectatorView.Unity/Assets/SpatialAlignment.ASA/Scripts/SpatialAlignment/SpatialAnchorsAndroidCoordinateService.cs [hunk 1]
```diff
@@ -30,16 +31,31 @@ public SpatialAnchorsAndroidCoordinateService(SpatialAnchorsConfiguration spatia
         /// <inheritdoc/>
         protected override Task OnInitializeAsync()
         {
+            if (initialized)
```
Comment (andreiborodin) [inline src/SpectatorView.Unity/Assets/SpatialAlignment.ASA/Scripts/SpatialAlignment/SpatialAnchorsAndroidCoordinateService.cs]: What I would recommend is to cache the taskCompletionSource in a static variable, and perform a null check on it to see if it needs instantiation or not. If it doesn't, just return it's task. Also, you can wrap all of it in a lock.

File: src/SpectatorView.Unity/Assets/SpatialAlignment.ASA/Scripts/SpatialAlignment/SpatialAnchorsCoordinateService.cs [hunk 2]
```diff
@@ -324,6 +324,11 @@ private async Task<ISpatialCoordinate> TryCreateCoordinateImplAsync(Vector3 worl
                     await session.CreateAnchorAsync(cloudSpatialAnchor);
                     return new SpatialAnchorsCoordinate(cloudSpatialAnchor, spawnedAnchorObject);
                 }
+                catch(TaskCanceledException)
```
Comment (andreiborodin) [inline src/SpectatorView.Unity/Assets/SpatialAlignment.ASA/Scripts/SpatialAlignment/SpatialAnchorsCoordinateService.cs]: I would recommend making this a OperationCancelledException, which is the base of TaskCancelledException

File: src/SpectatorView.Unity/Assets/SpatialAlignment.ASA/Scripts/SpectatorView/SpatialAnchorsLocalizer.cs [hunk 3]
```diff
@@ -90,17 +90,17 @@ private void Update()
             Updated?.Invoke();
         }
 
-        private class SpatialCoordinateLocalizationSession : DisposableBase, ISpatialLocalizationSession
+        private class SpatialCoordinateLocalizationSession : SpatialLocalizationSession
         {
-            public IPeerConnection Peer => peerConnection;
+            public override IPeerConnection Peer => peerConnection;
```
Comment (andreiborodin) [inline src/SpectatorView.Unity/Assets/SpatialAlignment.ASA/Scripts/SpectatorView/SpatialAnchorsLocalizer.cs]: Did you mean to add "///<inheritdoc/>" here as well?

File: src/SpectatorView.Unity/Assets/SpectatorView/Scripts/SpatialAlignment/MarkerVisual/MarkerVisualDetectorSpatialLocalizer.cs [hunk 4]
```diff
@@ -70,49 +71,59 @@ public LocalizationSession(MarkerVisualDetectorSpatialLocalizer localizer, Marke
...
             }
 
             /// <inheritdoc />
-            public async Task<ISpatialCoordinate> LocalizeAsync(CancellationToken cancellationToken)
+            public override async Task<ISpatialCoordinate> LocalizeAsync(CancellationToken cancellationToken)
             {
-                DebugLog($"Waiting for marker visual, CanBeCanceled:{cancellationToken.CanBeCanceled}, IsCancellationRequested:{cancellationToken.IsCancellationRequested}");
-                await Task.WhenAny(coordinateAssigned.Task, Task.Delay(-1, cancellationToken));
-                if (string.IsNullOrEmpty(coordinateId))
+                if (!defaultCTS.Token.CanBeCanceled)
```
Comment (andreiborodin) [inline src/SpectatorView.Unity/Assets/SpectatorView/Scripts/SpatialAlignment/MarkerVisual/MarkerVisualDetectorSpatialLocalizer.cs]: >defaultCTS.Token [](start = 21, length = 16)

Instead of exposing the CTS, I would recommend an auto property of type CancellationToken that you set with the CTS token as soon as you create it. This way, if the CTS is cancelled and disposed, the token can be still used in it's cancelled state instead of throwing objectdisposedexception

File: src/SpectatorView.Unity/Assets/SpectatorView/Scripts/SpatialAlignment/SpatialCoordinateSystemManager.cs [hunk 5]
```diff
@@ -178,7 +205,7 @@ public Task<bool> RunRemoteLocalizationAsync(SocketEndpoint socketEndpoint, Guid
             if (remoteLocalizationSessions.TryGetValue(socketEndpoint, out var currentCompletionSource))
             {
                 DebugLog("Canceling existing remote localization session: {socketEndpoint.Address}");
-                currentCompletionSource.TrySetCanceled();
+                currentCompletionSource.TrySetResult(false);
```
Comment (andreiborodin) [inline src/SpectatorView.Unity/Assets/SpectatorView/Scripts/SpatialAlignment/SpatialCoordinateSystemManager.cs]: >TrySetResult [](start = 40, length = 12)

Why not cancel here?

Comment (chrisfromwork) [reply src/SpectatorView.Unity/Assets/SpectatorView/Scripts/SpatialAlignment/SpatialCoordinateSystemManager.cs]: No preference. I guess I thought this would be called from a user action, but I will just revert.

File: src/SpectatorView.Unity/Assets/SpectatorView/Scripts/SpatialAlignment/WorldAnchor/WorldAnchorSpatialLocalizer.cs [hunk 6]
```diff
@@ -63,47 +63,62 @@ public override bool TryCreateLocalizationSession(IPeerConnection peerConnection
...
-                }
+
+                return await Task.FromResult(spatialCoordinate);
             }
 
-            public void OnDataReceived(BinaryReader reader)
+            /// <inheritdoc />
+            public override void OnDataReceived(BinaryReader reader)
             {
             }
```
Comment (matthejo) [inline src/SpectatorView.Unity/Assets/SpectatorView/Scripts/SpatialAlignment/WorldAnchor/WorldAnchorSpatialLocalizer.cs]: If this is now virtual and we don't need to explicitly not have the base method happen, we can just get rid of this empty method now

File: src/SpectatorView.Unity/Assets/SpectatorView/Scripts/SpatialAlignment/PhysicalMarker/MarkerDetectorSpatialLocalizer.cs [hunk 7]
```diff
@@ -46,26 +47,34 @@ public LocalizationSession(MarkerDetectorSpatialLocalizer localizer, MarkerDetec
...
                 }
 
                 return spatialCoordinate;
             }
 
-            public void OnDataReceived(BinaryReader reader)
+            /// <inheritdoc />
+            public override void OnDataReceived(BinaryReader reader)
             {
             }
```
Comment (matthejo) [inline src/SpectatorView.Unity/Assets/SpectatorView/Scripts/SpatialAlignment/PhysicalMarker/MarkerDetectorSpatialLocalizer.cs]: Can get rid of this override

File: src/SpectatorView.Unity/Assets/SpectatorView/Scripts/SpatialAlignment/SpatialLocalizationSession.cs [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+            if (!this.IsDisposed &&
+                defaultCTS != null &&
+                defaultCTS.Token.CanBeCanceled)
+            {
+                defaultCTS.Cancel();
+            }
+        }
+
+        /// <inheritdoc />
+        public abstract void OnDataReceived(BinaryReader reader);
```
Comment (matthejo) [inline src/SpectatorView.Unity/Assets/SpectatorView/Scripts/SpatialAlignment/SpatialLocalizationSession.cs]: >public abstract void OnDataReceived(BinaryReader reader); [](start = 8, length = 57)

I would just make this virtual with an empty body, since it's sometimes not needed by subclasses

File: src/SpectatorView.Unity/Assets/SpectatorView/Scripts/SpatialAlignment/SpatialCoordinateSystemParticipant.cs [hunk 9]
```diff
@@ -27,7 +27,7 @@ public class SpatialCoordinateSystemParticipant : DisposableBase, IPeerConnectio
         private bool showDebugVisuals;
         private readonly TaskCompletionSource<ISet<Guid>> peerSupportedLocalizersTaskSource = new TaskCompletionSource<ISet<Guid>>();
 
-        public SocketEndpoint SocketEndpoint { get; }
+        public SocketEndpoint SocketEndpoint { get; set; }
```
Comment (matthejo) [inline src/SpectatorView.Unity/Assets/SpectatorView/Scripts/SpatialAlignment/SpatialCoordinateSystemParticipant.cs]: >public SocketEndpoint SocketEndpoint { get; set; } [](start = 8, length = 50)

Under what circumstances can the endpoint associated with a participant change?

File: src/SpectatorView.Unity/Assets/SpectatorView/Scripts/SpatialAlignment/SpatialCoordinateSystemManager.cs [hunk 10]
```diff
@@ -265,16 +276,17 @@ private void SendLocalizationCompleteCommand(SocketEndpoint socketEndpoint, bool
 
         private void OnConnected(SocketEndpoint endpoint)
         {
-            if (participants.ContainsKey(endpoint))
+            if (participants.TryGetValue(endpoint.Address, out var existingParticipant))
```
Comment (matthejo) [inline src/SpectatorView.Unity/Assets/SpectatorView/Scripts/SpatialAlignment/SpatialCoordinateSystemManager.cs]: >if (participants.TryGetValue(endpoint.Address, out var existingParticipant)) [](start = 12, length = 76)

Why the change to use the address? I'm confused about when this could happen and why it's not an error.

File: src/SpectatorView.Unity/Assets/SpectatorView/Scripts/SpatialAlignment/SpatialCoordinateSystemParticipant.cs (hunk 9 above)
Comment (chrisfromwork) [reply src/SpectatorView.Unity/Assets/SpectatorView/Scripts/SpatialAlignment/SpatialCoordinateSystemParticipant.cs]: we saw locally that Unity buttons ended up having two event calls on a single press based on some input configurations. This caused two socket endpoints to get created for the same ip address. TCPConnectionManager has been updated to prevent creating multiple end points with subsequent event calls, but other networking abstractions may not be as friendly.

Do you believe it is safe to key off ip addresses or can you imagine scenarios where the same ip address creates multiple endpoints for the same shared experience?

Comment (matthejo) [reply src/SpectatorView.Unity/Assets/SpectatorView/Scripts/SpatialAlignment/SpatialCoordinateSystemParticipant.cs]: I could certainly imagine scenarios where this would break, e.g. if we wanted two devices both connected over IPOverUSB to work (on different ports). They would both have 127.0.0.1 as their address but be bound to different ports and have different endpoints.

---
In reply to: [337232321](https://github.com/microsoft/MixedReality-SpectatorView/pull/247#discussion_r337232321) [](ancestors = 337232321)

File: src/SpectatorView.Unity/Assets/SpatialAlignment/Scripts/Utilities/Extensions.cs [hunk 11]
```diff
@@ -182,9 +182,9 @@ public static Task AsTask(this CancellationToken cancellationToken)
         /// <param name="task">The task to await.</param>
         /// <param name="cancellationToken">The cancellation token to stop awaiting.</param>
         /// <returns>The task that can be awaited unless the cancellation token is triggered.</returns>
-        public static Task Unless(this Task task, CancellationToken cancellationToken)
+        public static async Task Unless(this Task task, CancellationToken cancellationToken)
         {
-            return Task.WhenAny(task, cancellationToken.AsTask());
+            await (await Task.WhenAny(task, cancellationToken.AsTask()));
```
Comment (andreiborodin) [inline src/SpectatorView.Unity/Assets/SpatialAlignment/Scripts/Utilities/Extensions.cs]: > await (await Task.WhenAny(task, cancellationToken.AsTask())); [](start = 11, length = 62)

Actually this could be just one await and teh function not async, but the bug was that there was no await at all in old version.

File: src/SpectatorView.Unity/Assets/SpectatorView/Scripts/SpatialAlignment/SpatialCoordinateSystemManager.cs [hunk 12]
```diff
@@ -48,11 +48,27 @@ public class SpatialCoordinateSystemManager : Singleton<SpatialCoordinateSystemM
         private const string LocalizationCompleteCommand = "LOCALIZEDONE";
         private readonly Dictionary<Guid, ISpatialLocalizer> localizers = new Dictionary<Guid, ISpatialLocalizer>();
         private readonly Dictionary<SocketEndpoint, TaskCompletionSource<bool>> remoteLocalizationSessions = new Dictionary<SocketEndpoint, TaskCompletionSource<bool>>();
-        private Dictionary<SocketEndpoint, SpatialCoordinateSystemParticipant> participants = new Dictionary<SocketEndpoint, SpatialCoordinateSystemParticipant>();
+        private Dictionary<string, SpatialCoordinateSystemParticipant> participants = new Dictionary<string, SpatialCoordinateSystemParticipant>();
         private HashSet<INetworkManager> networkManagers = new HashSet<INetworkManager>();
-        private ISpatialLocalizationSession currentLocalizationSession = null;
         private ITrackingObserver trackingObserver = null;
 
+        private object localizationLock = new object();
```
Comment (matthejo) [inline src/SpectatorView.Unity/Assets/SpectatorView/Scripts/SpatialAlignment/SpatialCoordinateSystemManager.cs]: >private [](start = 8, length = 7)

this field could also be readonly

File: src/SpectatorView.Unity/Assets/SpatialAlignment.ASA/Scripts/SpatialAlignment/SpatialAnchorsCoordinateService.cs [hunk 13]
```diff
@@ -334,6 +334,8 @@ private async Task<ISpatialCoordinate> TryCreateCoordinateImplAsync(Vector3 worl
             {
                 ReleaseSessionStartRequest();
             }
+
+            return null;
```
Comment (matthejo) [inline src/SpectatorView.Unity/Assets/SpatialAlignment.ASA/Scripts/SpatialAlignment/SpatialAnchorsCoordinateService.cs]: >return null; [](start = 12, length = 12)

This seems like unreachable code, how would this ever be hit?

File: src/SpectatorView.Unity/Assets/SpectatorView/Scripts/SpatialAlignment/SpatialCoordinateSystemParticipant.cs (hunk 9 above)
Comment (chrisfromwork) [reply src/SpectatorView.Unity/Assets/SpectatorView/Scripts/SpatialAlignment/SpatialCoordinateSystemParticipant.cs]: i will revert this now.

File: src/SpectatorView.Unity/Assets/SpatialAlignment/Scripts/Utilities/Extensions.cs (hunk 11 above)
Comment (chrisfromwork) [reply src/SpectatorView.Unity/Assets/SpatialAlignment/Scripts/Utilities/Extensions.cs]: we tested this together locally and will leave this how it is.

File: src/SpectatorView.Unity/Assets/SpectatorView/Scripts/SpatialAlignment/SpatialCoordinateSystemManager.cs (hunk 10 above)
Comment (chrisfromwork) [reply src/SpectatorView.Unity/Assets/SpectatorView/Scripts/SpatialAlignment/SpatialCoordinateSystemManager.cs]: reverting

File: src/SpectatorView.Unity/Assets/SpatialAlignment.ASA/Scripts/SpatialAlignment/SpatialAnchorsCoordinateService.cs (hunk 13 above)
Comment (chrisfromwork) [reply src/SpectatorView.Unity/Assets/SpatialAlignment.ASA/Scripts/SpatialAlignment/SpatialAnchorsCoordinateService.cs]: product of removing a catch stub, fixing now.

File: src/SpectatorView.Unity/Assets/SpectatorView/Scripts/SpatialAlignment/SpatialCoordinateSystemManager.cs [hunk 14]
```diff
@@ -50,9 +50,25 @@ public class SpatialCoordinateSystemManager : Singleton<SpatialCoordinateSystemM
         private readonly Dictionary<SocketEndpoint, TaskCompletionSource<bool>> remoteLocalizationSessions = new Dictionary<SocketEndpoint, TaskCompletionSource<bool>>();
         private Dictionary<SocketEndpoint, SpatialCoordinateSystemParticipant> participants = new Dictionary<SocketEndpoint, SpatialCoordinateSystemParticipant>();
         private HashSet<INetworkManager> networkManagers = new HashSet<INetworkManager>();
-        private ISpatialLocalizationSession currentLocalizationSession = null;
         private ITrackingObserver trackingObserver = null;
 
+        private readonly object localizationLock = new object();
```
Comment (stephenatwork) [inline src/SpectatorView.Unity/Assets/SpectatorView/Scripts/SpatialAlignment/SpatialCoordinateSystemManager.cs]: Is there a reason to prefer a separate lock object? If this is only protecting currentLocalizationSession, would we lock that object directly instead?

Comment (chrisfromwork) [reply src/SpectatorView.Unity/Assets/SpectatorView/Scripts/SpatialAlignment/SpatialCoordinateSystemManager.cs]: locking on a null object throws exceptions. Since we want to populate currentLocalizationSession within the lock, it seems safer to lock on a separate, non-null object.
